velox/expression/EvalCtx.h

```diff
@@ -32,6 +34,180 @@ class LocalSelectivityVector;
 struct ContextSaver;
 class PeeledEncoding;
 
+/// Tracks per-row errors that occurred during expression evaluation.
+/// Used when EvalCtx::throwOnError() is false.
+class EvalErrors {
+ public:
+  EvalErrors(memory::MemoryPool* pool, vector_size_t size) {
+    // Do not allocate 'values' buffer. It uses ~20 bytes per row and it may not
+    // be needed.
+    errors_ = std::make_shared<ErrorVector>(
+        pool,
+        OpaqueType::create<void>(),
+        allocateNulls(size, pool, bits::kNull),
+        size,
+        nullptr,
+        std::vector<BufferPtr>{});
+  }
+
+  vector_size_t size() const {
+    return errors_->size();
+  }
+
+  /// Similar to std::vector::reserve. Allocates internal buffers to fit at
+  /// least 'size' rows. No-op if 'size()' is already at or exceeding requested.
+  void ensureCapacity(vector_size_t size) {
+    if (errors_->size() >= size) {
+      return;
+    }
+
+    const auto oldSize = errors_->size();
+    errors_->resize(size, false);
+    // Set all new positions to null, including the one to be set.
+    for (auto i = oldSize; i < size; ++i) {
+      errors_->setNull(i, true);
+    }
+  }
+
+  /// Returns true if at least one row has an error.
+  bool hasError() const {
+    const auto firstErrorRow =
+        bits::findFirstBit(errors_->rawNulls(), 0, errors_->size());
+    return firstErrorRow >= 0;
+  }
+
+  /// Returns true if 'index' has an error.
+  bool hasErrorAt(vector_size_t index) const {
+    return index < errors_->size() && !errors_->isNullAt(index);
+  }
+
+  /// Throws if 'index' has an error. The caller must ensure that error details
+  /// are available.
+  void throwIfErrorAt(vector_size_t index) const {
+    if (hasErrorAt(index)) {
+      auto error = errors_->valueAt(index);
+      VELOX_CHECK_NOT_NULL(error);
+      std::rethrow_exception(
+          *std::static_pointer_cast<std::exception_ptr>(error));
+    }
+  }
+
+  /// Finds first row in 'rows' that has an error and throws that error. The
+  /// caller must ensure that error details are available.
+  void throwFirstError(const SelectivityVector& rows) const {
+    const auto errorSize = errors_->size();
+    rows.testSelected([&](vector_size_t row) {
+      if (row < errorSize) {
+        throwIfErrorAt(row);
+        return true;
+      }
+      return false;
+    });
+  }
+
+  /// Returns std::nullopt if 'index' doesn't have an error.
+  /// Returns nullptr if 'index' has an error, but error details are not
+  /// available. Returns std::exception_ptr if 'index' has an error and error
+  /// details are available.
+  std::optional<std::shared_ptr<std::exception_ptr>> errorAt(
```

Comment:
nit: this returning std::nullptr vs nullptr is a bit subtle, wondering if we should return something like a std::pair<isError, ExceptionPtr> instead? This would make is explicit whether there is an error